src/common/telemetry/src/tracing_sampler.rs

```diff
@@ -0,0 +1,205 @@
+impl PartialEq for TracingSampleOptions {
+    fn eq(&self, other: &Self) -> bool {
+        self.default_ratio == other.default_ratio && self.rules == other.rules
+    }
+}
+impl PartialEq for TracingSampleRule {
+    fn eq(&self, other: &Self) -> bool {
+        self.protocol == other.protocol
+            && self.request_types == other.request_types
+            && self.ratio == other.ratio
+    }
+}
+
+impl Eq for TracingSampleOptions {}
+impl Eq for TracingSampleRule {}
+
+pub fn create_sampler(opt: &TracingSampleOptions) -> Box<dyn ShouldSample> {
+    if opt.rules.is_empty() {
+        Box::new(Sampler::TraceIdRatioBased(opt.default_ratio))
+    } else {
+        Box::new(RuleSampler::new(opt))
+    }
+}
+
+/// Determine the sampling rate of a span according to the `rules` provided in `RuleSampler`.
+/// For spans that do not hit any `rules`, the `default_ratio` is used.
+#[derive(Debug, Clone)]
+struct RuleSampler {
+    default_ratio: f64,
+    rules: Vec<Rule>,
+}
+
+#[derive(Debug, Clone)]
+struct Rule {
+    protocol: String,
+    request_types: HashSet<String>,
+    ratio: f64,
+}
+
+impl Rule {
+    pub fn new(rule: &TracingSampleRule) -> Self {
+        Self {
+            protocol: rule.protocol.clone(),
+            request_types: rule.request_types.iter().cloned().collect(),
+            ratio: rule.ratio,
+        }
+    }
+
+    pub fn match_rule(&self, protocol: &str, request_type: Option<&str>) -> Option<f64> {
+        if protocol == self.protocol {
+            if self.request_types.is_empty() {
+                Some(self.ratio)
+            } else if request_type.is_none() {
+                None
+            } else if self.request_types.contains(request_type.unwrap()) {
+                Some(self.ratio)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+impl RuleSampler {
+    pub fn new(opt: &TracingSampleOptions) -> Self {
+        Self {
+            default_ratio: opt.default_ratio,
+            rules: opt.rules.iter().map(Rule::new).collect(),
+        }
+    }
+}
+
+impl ShouldSample for RuleSampler {
+    fn should_sample(
+        &self,
+        parent_context: Option<&opentelemetry::Context>,
+        trace_id: TraceId,
+        _name: &str,
+        _span_kind: &SpanKind,
+        attributes: &[KeyValue],
+        _links: &[Link],
+    ) -> SamplingResult {
+        let (mut protocol, mut request_type) = (None, None);
+        for kv in attributes {
+            match kv.key.as_str() {
+                "protocol" => protocol = Some(kv.value.as_str()),
+                "request_type" => request_type = Some(kv.value.as_str()),
+                _ => (),
+            }
+        }
+        let ratio = protocol
+            .map(|p| {
+                self.rules
+                    .iter()
+                    .find_map(|rule| rule.match_rule(p.as_ref(), request_type.as_deref()))
+                    .unwrap_or(self.default_ratio)
+            })
```

Comment:
Seems you can use `and_then`:
```suggestion
            .and_then(|p| {
                self.rules
                    .iter()
                    .find_map(|rule| rule.match_rule(p.as_ref(), request_type.as_deref()))
            })
```